DESIGN.md

```diff
@@ -0,0 +1,247 @@
+# Linux Worker Service using Namespaces/CGroups Implemented in Golang
+
+This is an experiement to build a job worker service that provides an API to run arbitrary Linux processes.
+
+There will be 3 components to make up this service:
+
+1. A Golang library to start, stop, and query the status/output of a worker job
+2. A gRPC API to expose the functionality of this library over a network to start, stop, query status and stream output of a job, including authentication via mTLS and authorization scheme
+3. A CLI for users to interact with the API to start, stop, query status/stream output of a job
+
+> :pencil: Note that code and command samples here are meant to be approximations of the final implementation. Slight variations may exist in what is actually implemented in code from what you see in this design.
+
+## Library
+
+This will be the lowest-level part of the service stack, with the following goals and outcomes:
+
+* providing an interface for a linux command/process to be executed on behalf of a user/owner
+* interfacing with the linux OS internals/kernel via [`sys/unix`](https://pkg.go.dev/golang.org/x/sys/unix) to:
+  * Add process isolation for a worker process in the following ways:
+    * **PID**: no other processes executed by the worker service can see another worker process
+    * **mounts**: dedicate file system space for the process
+    * **networking**: isolate network interface and connectivity out of the linux machine, but with no routes to other processes/namespaces
+  * Add resource usage control by assigning the spawned process to a cgroup, and limiting the amount of:
+    * **CPU used by the process**: hard limit on the total CPU, **1 CPU**, that any single process could use from the host machine
+    * **Memory used by the process**: hard limit on the total memory, **100 MB**, that any single process could use from the host machine
+    * **Disk I/O rate used by the process**: hard limit on the read/write bytes/second available, **1 MB/s** to any single process
+* Returning a unique service process ID for a successfully-triggered process/worker
+* Ability to stop a running process by using this unique process ID
+* Ability to get/stream the output/logs of the worker process by using the process ID, as well as get the final result exit code when available
+
+### Library-Defined Process Store Interface
+
+The library will define an interface for storing and tracking running processes, a process store, like:
+
+```golang
+// ProcessStore defines an interface for defining custom stores to track processes, ownership, status, output, etc.
+type ProcessStore interface {
+  NewProcess(workerID string, ownerID string) (err error)
+  SetHostPID(workerID string, hostPID string) (err error)
+  GetHostPID(workerID string) (hostPID string, err error)
+  GetOutputWriters(workerID string) (stdout io.Writer, stderr io.Writer, err error)
+  GetOwnerID(workerID string) (ownerID string, err error)
```

Comment:
Does it make sense to return a ownerID in GetStatus in order to reduce the interface complexity ?

Reply:
https://github.com/rockholla/go-linux-ns-cgroup-worker/pull/1/commits/05077ce4decc133bb758e2b540415a5e6136ad8b and https://github.com/rockholla/go-linux-ns-cgroup-worker/pull/1/commits/f969dff477f9609edd7e133a91a728a9e53d5267